### DISCUSSION TITLE
Can't implement/authenticate Custom Provider for Airflow using FAB

### DISCUSSION THREAD

**Discussion by h00jraq** [apache/airflow#31267] [Q&A]
Hello,
I'm quite new in FAB and in Authentication, so please forgive me if I made any stupid mistake here.

### Environment
Apache Airflow 2.5.0 deployed via helm chart on AKS Cluster
Flask-Appbuilder version: 4.3.0

I'm trying to authorize using custom provider, similar to what was done here: 
https://github.com/dpgaspar/Flask-AppBuilder/issues/608#issuecomment-845264649
I was trying to follow that approach but without success.

I have 3 problems for now:

1. I can't login to Airflow - I've pasted logs at the bottom. Provider is initalized with correct provider name, I can see that valid request is send to the identity server but then it fails with message ` Error authorizing OAuth access token: invalid_client: Client authentication failed`. Thing is, Client Id/secret are 100% correct but maybe they are not taken from remote app? Could not print them in oauth_user_info, even if I tried, nothing was showing in webserver logs.

2. Unfortunately, it looks like none of the method (except __init__ if I add one) is called from class pointed out by FAB_SECURITY_MANAGER_CLASS. I have added few print statements and they are just not in the logs. 
I was trying to `use get_oauth_user_info()` but result was the same, no print statements and ended with same errors:

3. I can see that class is used because all values from my custom provider are used except redirect URI, which is always pointing to url/oauth-authorized/provider-name and I can't override that. I see that there is @expore(/oauth-authorized/provider) in the code but I thought that my redirect_uri should be used.

```python
#!/usr/bin/python
# -*- coding: utf-8 -*-
import os
from airflow.configuration import conf
from airflow.utils.log.logging_mixin import LoggingMixin
from flask_appbuilder.security.manager import AUTH_OAUTH
from airflow.www.security import AirflowSecurityManager

SQLALCHEMY_DATABASE_URI = conf.get('core', 'SQL_ALCHEMY_CONN')
basedir = os.path.abspath(os.path.dirname(__file__))
CSRF_ENABLED = True
AUTH_TYPE = AUTH_OAUTH
AUTH_USER_REGISTRATION_ROLE = 'Public'
AUTH_USER_REGISTRATION = True


class CustomSecurityManager(AirflowSecurityManager):

    def oauth_user_info(self, provider, response=None):
        sm = self.appbuilder.sm
        if provider == 'pwc':
            response = sm.oauth_remotes[provider].get('info')
            data = response.json()
            self.log.debug('PWC response received : {0}'.format(provider))

            # remote_app = self.appbuilder.sm.oauth_remotes[provider]

            print 'something'
            print remote_app

            # me = remote_app.get("userinfo_endpoint")
            # me = self.appbuilder.sm.oauth_remotes[provider].get("userinfo_endpoint")
            # id_token = response.get("id_token")

            logging.error(me)
            return {
                'username': data['username'],
                'email': data.get('email', ''),
                'first_name': data.get('first_name', ''),
                'last_name': data.get('last_name', ''),
                }
        else:

            return {}

    def sync_roles(self):

        # Custom roles are given airflow website permission in AirflowSecurityManager.sync_roles()

        self.add_role('Guest')
        return super().sync_roles()


# a mapping from the values of `userinfo["role_keys"]` to a list of FAB roles

AUTH_ROLES_MAPPING = {'airflow_nonprod_admin': ['Admin'],
                      'airflow_viewer': ['Public']}

AUTH_ROLES_SYNC_AT_LOGIN = True

OAUTH_PROVIDERS = [{
    'name': 'my-provider',
    'token_key': 'access_token',
    'icon': 'fa-globe',
    'remote_app': {
        'client_id': 'urn:insightshub-test.pwcinternal.co.uk',
        'client_secret': 'xxxxxx',
        'scope': 'openid email profile',
        'issuer': 'https://example.com.com:443/openam/oauth2',
        'token_endpoint': 'https://example.com.com/openam/oauth2/access_token'
            ,
        'userinfo_endpoint': 'https://example.com/openam/oauth2/userinfo'
            ,
        'redirect_uri': 'https://example.com/oauth/callback',
        'access_token_url': 'https://example.com/openam/oauth2/access_token'
            ,
        'access_token_params': {'scope': 'openid email profile'},
        'authorize_url': 'https://lexample.com/openam/oauth2/authorize'
            ,
        'authorize_params': {'scope': 'openid email profile'},
        'jwks_uri': 'https://example.com/openam/oauth2/connect/jwk_uri'
            ,
        },
    }]

FAB_SECURITY_MANAGER_CLASS = 'webserver_config.CustomSecurityManager'
```

I'm redirected to my provider login page, trying to log with sso and it works but then I'm taken to the Airflow login page with message "The request to sign in was denied". Below logs are taken from Airflow-webserver. Problem is, client id/secret are 100% correct but maybe they are not taken from remote_app but from other place and are just ignored same as redirect_uri?

```pytb
[2023-05-12T21:26:40.079+0000] {views.py:651} DEBUG - Authorized init
[2023-05-12T21:26:40.079+0000] {views.py:651} DEBUG - Authorized init
[2023-05-12T21:26:40.083+0000] {connectionpool.py:1007} DEBUG - Starting new HTTPS connection (1): [example.com:443](http://example.com:443/)
[2023-05-12T21:26:40.280+0000] {connectionpool.py:465} DEBUG - [https://examplecom:443](https://examplecom/) "POST /openam/oauth2/access_token HTTP/1.1" 401 77
[2023-05-12T21:26:40.284+0000] {views.py:659} ERROR - Error authorizing OAuth access token: invalid_client: Client authentication failed
[2023-05-12T21:26:40.284+0000] {views.py:659} ERROR - Error authorizing OAuth access token: invalid_client: Client authentication failed
```

**Comment by potiuk** [comment]
I do not think it is going to work. Airlfow uses hevily trimmed down version of SecurityManager from FAB due to some problems we had in the past and Airlfow's security manager is not supporting replacing it with a custom one.

In fact we are even about to vote the new AIP-56 where we move away from FAB (is one of the reasons that we are not experts in this parts) into a more isolated and separated approach where authentication and authorisation will be handled differently)
 https://cwiki.apache.org/confluence/display/AIRFLOW/AIP-56+Extensible+user+management 

If you need something that out-of-the-box is not supported by Airlfow, my recommendation is to do the same on its own now. Rely on external authentication (keycloak for example) and do all authantication there, then configure airflow to be behind keycloak and accept pre-authenticated headers from it  - this basically what medium/long term will be the default way of authenticating with Airflow. I know it can be done (not sure exactly how) and we are planning to do it.

**Reply by h00jraq** [reply]
The thing is, we cant use any other Identity /access management solution as this is the only one approved in the company (custom provider hosted on apenAm).

I really hoped that by overriding some methods I can get it to work as for example I saw someone whas using OIDC with gitlab provider and it worked.

IT complains about the token So that may be an issue  but I Don know how to check it.

**Reply by potiuk** [reply]
Talk to FAB,

**Comment by expore** [comment]
您好，我已收到您的邮件，待阅读后会尽快给您回复。
